### PR TITLE
multiroom envs

### PR DESCRIPTION
**Pull request by sriyash421**
Initial Structure almost similar to the python version. Would love suggestions on how to use additional Julia functionality to make it more efficient.

Currently adopts a recursive brute force method to generate rooms.

### CODE REVIEW

**Comment by findmyway** [comment]
@sriram13m Could you help to review this PR?

@sriyash421 Could you also add a GIF and update the README?

File: src/envs/multiroom.jl [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    return true
+end
+
+function MultiRoom(;n=25, minRooms, maxRooms, maxSize=10, rng=Random.GLOBAL_RNG)
+    objects = (EMPTY, WALL, GOAL, (Door(c) for c in COLORS)...)
+    world = GridWorldBase(objects, n, n)
+    world[EMPTY, :, :] .= true
+    numRooms = rand(rng, minRooms:maxRooms)
+    roomList = []
+    while length(roomList) < numRooms:
```
Comment (Sid-Bhatia-0) [inline src/envs/multiroom.jl:78]: @sriyash421 , I was wondering if you share a working example with environment. There seems to be a `:` at the end of line 78, which won't allow the code to compile.

**Comment by Sid-Bhatia-0** [comment]
@sriyash421 , maybe consider making this PR as a draft?

There might be several other changes to be made including
1. Formatting variables in lower case (as per Julia style guide [here](https://docs.julialang.org/en/v1/manual/variables/#Stylistic-Conventions)).
2. Also `placeRooms` seems to be a function with side-effects. So it would be helpful to rename it to `placeRooms!`, for example.
3. Adding GIF and updating README as suggested by @findmyway 
4. Include the environment in `env.jl` and test it.

I am wondering about the goal of this environment and if we can rethink its design to make it better. A few questions that come to my mind:

1. Why does the gym-minigrid version have doors without keys? Instead, could we just replace the doors with empties.
2. Is the concept of a multi-room translationally invariant within the grid? Maybe we can first construct the multi-room iteratively and then place it within the grid so that it fits?

Would love to hear other peoples' thoughts on this.

File: src/envs/multiroom.jl (hunk 1 above)
Comment (sriyash421) [reply src/envs/multiroom.jl:78]: Yeah. Sorry I haven't pushed all the changes yet. Will do that after I test it finally.

**Comment by sriyash421** [comment]
> @sriyash421 , maybe consider making this PR as a draft? (By the way, it'd be nice if someone could enlighten me on the difference between a [WIP] PR vs. a draft PR :p )
> 
> There might be several other changes that could be made including, but not limited to the following:
> 
> 1. Formatting variables in lower case (as per Julia style guide [here](https://docs.julialang.org/en/v1/manual/variables/#Stylistic-Conventions)).
> 2. Also `placeRooms` seems to be a function with side-effects. So it would be helpful to rename it to `placeRooms!`, for example.
> 3. Adding GIF and updating README as suggested by @findmyway
> 4. Include the environment in `env.jl` and test it.
> 5. You can also add default arguments for the environment so that something like `w = MultiRoom()` just works out of the box. I feel it helps one get started with playing with the environmnet more quickly.
> 
> I am wondering about the goal of this environment and if we can rethink its design to make it better. A few questions that come to my mind:
> 
> 1. Why does the gym-minigrid version have doors without keys? Instead, could we just replace the doors with empties.
> 2. Is the concept of a multi-room translationally invariant within the grid? Maybe we can first construct the multi-room iteratively and then place it within the grid so that it fits?
> 
> Would love to hear other peoples' thoughts on this.

I have added some the features you mentioned here, and will push a complete version in sometime.

**Comment by sriyash421** [comment]
@findmyway @Sid-Bhatia-0 
1. Added MultiRoomNXSY envs
2. Added a list to sample directions

**Comment by sriyash421** [comment]
> @sriyash421 , maybe consider making this PR as a draft? (By the way, it'd be nice if someone could enlighten me on the difference between a [WIP] PR vs. a draft PR :p )
> 
> There might be several other changes that could be made including, but not limited to the following:
> 
> 1. Formatting variables in lower case (as per Julia style guide [here](https://docs.julialang.org/en/v1/manual/variables/#Stylistic-Conventions)).
> 2. Also `placeRooms` seems to be a function with side-effects. So it would be helpful to rename it to `placeRooms!`, for example.
> 3. Adding GIF and updating README as suggested by @findmyway
> 4. Include the environment in `env.jl` and test it.
> 5. You can also add default arguments for the environment so that something like `w = MultiRoom()` just works out of the box. I feel it helps one get started with playing with the environmnet more quickly.
> 
> I am wondering about the goal of this environment and if we can rethink its design to make it better. A few questions that come to my mind:
> 
> 1. Why does the gym-minigrid version have doors without keys? Instead, could we just replace the doors with empties.
> 2. Is the concept of a multi-room translationally invariant within the grid? Maybe we can first construct the multi-room iteratively and then place it within the grid so that it fits?
> 
> Would love to hear other peoples' thoughts on this.

MiniGrid has different colored doors probably to differentiate between the entry and exit for certain rooms maybe. Else in case of partial observation, there may arise situations where they give the same state from certain coordinates. @findmyway Is this a reasonable explanation?

@Sid-Bhatia-0 @findmyway

**Comment by sriyash421** [comment]
`
qERROR: failed process: Process(`/home/sriyash/.julia/artifacts/7f40eeb66d90d3026ae5fb68761c263b57adb840/bin/ffmpeg -loglevel quiet -i '/tmp/jl_1uYe3U/##video#270.mkv' -i /tmp/jl_1uYe3U/palette.bmp -lavfi 'fps=24,scale=iw:ih:flags=lanczos [x]; [x][1:v] paletteuse' -y MultiRoom.gif`, ProcessExited(1)) [1]

`

Generating the gif throws this error

**Comment by sriyash421** [comment]
@findmyway Can you tell me why the CI/CD checks haven't performed for this PR? I cannot figure out why?

**Comment by findmyway** [comment]
See https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/issues/151

**Comment by sriyash421** [comment]
@findmyway I think rebase has messed up the commit history. So you can still merge this right?

**Comment by Sid-Bhatia-0** [comment]
@sriyash421 I'll help review the PR once I understand the environment completely.
By the way, CI has been fixed in #80

File: src/envs/multiroom.jl [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+using Random
+
+mutable struct MultiRoom{W<:GridWorldBase, R} <: AbstractGridWorld
+    world::W
+    agent_pos::CartesianIndex{2}
+    agent::Agent
+    goal_reward::Float64
+    reward::Float64
+    roomlist::AbstractArray
```
Comment (Sid-Bhatia-0) [inline src/envs/multiroom.jl:11]: I think this can be made `Array{Room}`

File: src/envs/multiroom.jl [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        success = place_rooms!(n, numleft-1, roomlist, minsize,  maxsize, next_entrywall, exit_doorpos)
+        if success 
+            break 
+        end
+    end
+    return true
+end
+
+function MultiRoom(;n=25, minrooms=2, maxrooms=4, maxsize=10, rng=Random.GLOBAL_RNG)
+    objects = (EMPTY, WALL, GOAL, (Door(c) for c in COLORS)...)
```
Comment (Sid-Bhatia-0) [inline src/envs/multiroom.jl:80]: I don't think we are utilizing all the `Door`s. We can use only as many as needed.

File: src/envs/multiroom.jl [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    while length(roomlist) < numrooms
+        cur_roomlist = []
+        entry_doorpos = CartesianIndex(rand(rng, 1:n-2),rand(rng, 1:n-2))
+        place_rooms!(n, numrooms, cur_roomlist, 4, maxsize, 2, entry_doorpos)
+        if length(cur_roomlist) > length(roomlist)
+            roomlist = cur_roomlist
+        end
+    end
+
+    door_colors = COLORS[randperm(rng, length(COLORS))][1:length(roomlist)-1]
```
Comment (Sid-Bhatia-0) [inline src/envs/multiroom.jl:94]: You can use `end` to refer to `length(roomlist)` everywhere for indexing.

**Comment by Sid-Bhatia-0** [comment]
The `Door` is not getting rendered properly. For your convenience, I have identified it to be happening right after using the `reset!` method.

Screenshot from 2020-11-24 20-11-44

Also, the latest master has CI. So you can rebase to pass the tests on that. We usually squash and merge so maybe the present history is okay, although a clean history is always better. @findmyway can guide better here.

My hunch is that we can do a simpler implementation for `MultiRoom` than the gym-minigrid one. I still need time to flesh out my thoughts to see if I can come up with something. Anyways, refactoring this one would infeasible right now via such a back-and-forth review process. So I'll give it a shot in a separate PR.

**Comment by Sid-Bhatia-0** [comment]
#84

**Pull request closed without merging**